### DISCUSSION TITLE
Fix EJS security vulnerability

### DISCUSSION THREAD

**Discussion by PremekPan** [nrwl/nx#17068] [Feature Requests]
Hi Team,

BlackDuck is reporting a medium security risk for EJS ([CVE-2023-29827](https://nvd.nist.gov/vuln/detail/CVE-2023-29827))
EJS devs already acknowledged the problem.

https://github.com/mde/ejs/issues/720
https://github.com/mde/ejs/blob/main/SECURITY.md#out-of-scope-vulnerabilities

BlackDuck is very clear about the situation:

> Description
> ejs is vulnerable to server-side template injection (SSTI) leading to remote code execution (RCE) when an application using ejs has user controlled input passed directly into the res.render() function. This could allow a remote attacker to execute commands on the underlying web server of such an application.
> 
> Note: The validity of [this](https://github.com/mde/ejs/issues/720#issuecomment-1540435858) vulnerability is contested by the maintainer of this application in this comment, since they already have a specific guidance against passing user input to res.render() [here](https://github.com/mde/ejs/blob/main/SECURITY.md#out-of-scope-vulnerabilities).

I understand, that EJS is just a tool, and as long as clients (nx) use it properly all should be fine.

Still, though, the medium security issue is a problem for organizations with strict security rules (no highs and mediums). You have to do a lot of paperwork to report an exception. And they can always say that you never know how this library will be used in the future and reject it...

Would you consider switching templating solution to something different?

**Comment by PremekPan** [comment]
Related to: https://github.com/nrwl/nx/issues/17072

**Comment by kaydeejay** [comment]
I am in one of those hypothetical situations that @PremekPan has mentioned, in that I work for a company that wants to know why we want to use nx when it has this vulnerable sub-dependency. I will be asking for an exception, but I would like to know if it's possible to get any assurances that nx does not use ejs in a way that makes it susceptible to template injection. That way I can pass it along to our security auditors.

**Reply by AgentEnder** [reply]
EJS is only used when running generators, its not exploitable as its not something thats hit on a server somewhere. The EJS vulnerability refers to using it as a server side templating language where a user could potentially inject calls containing arbitrary JS that would run on the server. 

Imagine a blog that contained a ejs template for the page that was something like

```ejs
<div class="post">
  <%- include('user/show', {user: user}); %>
  <div class="comments">
    <% comment.forEach((comment) => { %>
      <%- include('fragments/comment', { comment }); %>
    <% }) %>
  </div>
</div>
```

In this scenario, if the comment submitted contained an EJS fragment that wasn't sanitized, you'd have a template injection attack. This is similar to a SQL injection attack that you may be more familiar with.

In Nx, EJS is invoked as the result of a CLI command and not during your application serve. Its running on your machine. For anyone to exploit the vulnerability, they'd already have to be able to run commands on your machine, so its kind of a moot point.
